#166272333 jwt successful registration

adex001 opened the pull request:
#### What does this PR do?
- Successfully generates a token on successful login and registration
#### Description of Task to be completed?
- Write a function to create a token from the payload supplied.
- Generate token when user signs in or register.
- Implement login functionality.
- Reroute the current endpoint to make use of `POST api/v1/auth/login.js`.
- Create a function to verify a token.
#### How should this be manually tested?
- Navigate to the URL: `POST /api/v1/auth/register` then, register a new user.
- Navigate to the URL: `POST /api/v1/auth/login` and login a user. Verify the token gets displayed.
#### Any background context you want to provide?
- N/A
#### What are the relevant pivotal tracker stories?
[#166272333](https://www.pivotaltracker.com/story/show/166272333)
#### Screenshots (if appropriate)
- N/A
#### Questions:
- N/A

Commit:
feat-[166272333]: Successful token for login and signup
- create a create-token class

adex001 committed:
feat-[166272333]: implement token upon signin or signup
- write a function to create token from payload
- generate token when user signin or signup
- implement the login functionality
- reroute from the current endpoint to include api/v1/auth/*.js
- create a function to verify token

adex001 committed:
feat-[166272333]: implement token upon signin or signup
- write a function to create token from payload
- generate token when user signin or signup
- implement the login functionality
- reroute from the current endpoint to include api/v1/auth/*.js
- create a function to verify token

fidelisojeah commented on server/helpers/Token.js:
```diff
@@ -0,0 +1,41 @@
+    const { id, email } = user;
+    const token = await jwt.sign({
+      userId: id,
+      email,
```
how long does this token take before expiring?

You could place this value as an environment variable (and give a default value if one is not supplied)

adex001 replied:
We concluded not to let the token expire for now so that we won't be doing a lot of changing token in development but feedback noted. Thanks!

fidelisojeah replied:
then you do not understand my comment then.
put the value of expiration as an environment variable (this way, it's not encoded and everyone can modify as needed

adex001 replied:
ok. I understand better. Thanks!

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -59,20 +62,22 @@ class UserController {
+          email,
+        }
+      });
+      if (!user) return res.status(400).json({ status: 400, error: 'Invalid email or password' });
```
wrong error format.
use:
```javascript
errors:{
  global: 'Invalid email or password'
}
```

fidelisojeah commented on server/helpers/Token.js:
```diff
@@ -0,0 +1,41 @@
+    if (typeof token === 'undefined') return res.status(400).json();
+    const validToken = await jwt.verify(token, process.env.SECRET);
+    if (!validToken) return res.status(401).json();
+    req.validToken = validToken;
```
could this be called something else?

`validToken` is not exactly what it is since the token has already been decoded and verified

adex001 replied:
I will go with 'decoded'. Thanks!

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -59,20 +62,22 @@ class UserController {
+      // verify password
+      const match = await bcrypt.compare(password, user.password);
+      if (!match) return res.status(400).json({ status: 400, error: 'Invalid email or password' });
+      const token = await Token.create(user);
```
You are creating a token with the entire user object?

tokens are not encrypted. they are just a `base64` encoding of the object and a signed checksum of the object.

What you have here exposes our users sensitive details like password (even though this is hashed, it's still a security risk).

You are okay (for now with encoding only the users `id` and `username`)

adex001 replied:
In the token class, I have extracted the id and email from the user. I am encrypting only the id and email. Is that okay?

fidelisojeah replied:
no.

You may have reason to reuse the `Token.create` method.

sending only what you want to encode to it would make it easier to reuse

adex001 replied:
ok. I understand

fidelisojeah commented on server/helpers/Token.js:
```diff
@@ -0,0 +1,41 @@
+   * @memberof Token
+   */
+  static async verifyToken(req, res, next) {
+    const { token } = req.headers;
```
please wrap in a `try/catch` block. the `jwt.verify` throws errors on failure

adex001 replied:
I have implemented this.

fidelisojeah commented:
Please write tests for what you are implementing. I think the ticket mentioned something about doing the token thing on sign up as well (I'm personally not a fan of such practice). 

You can decide to do that or not (it's up to you). If you decide you don't want such behaviour, then you could modify the sign up process to only return `User Created Successfully` and no user object

adex001 committed:
feat-[166272333]: implement token upon signin or signup
- merge conflict from pulling from develop
- implement feedback from TTL

adex001 committed:
feat-[166272333]: implement token upon signin or signup
- implement feedback from TTL

adex001 committed:
feat-[166272333]: implement token upon signin or signup
- implement feedback by updating from validtoken to verifiedtoken
- Update tests to reflect token

fidelisojeah commented on server/helpers/Token.js:
```diff
@@ -0,0 +1,61 @@
+   * @memberof Token
+   */
+  static async create(payload) {
+    const { id, email } = payload;
```
why is this still here?

fidelisojeah commented on server/helpers/Token.js:
```diff
@@ -0,0 +1,61 @@
+    try {
+      const { token } = req.headers;
+      if (typeof token === 'undefined') {
+        return res.status(400).json({
```
you could use your global error handler here.

adex001 replied:
This has been implemented.

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -60,20 +65,26 @@ class UserController {
-      }
-      return res.status(400).json(info);
-    })(req, res, next);
+    try {
```
A validator was created for input (see sign up) I expect that you make use of it.
You may need to modify the values sync with @despeauxz

adex001 replied:
It has been implemented.

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -60,20 +65,26 @@ class UserController {
+          email,
+        }
+      });
+      if (!user) return res.status(400).json({ status: 400, errors: { global: 'Invalid email or password' } });
```
Again, sync with @despeauxz on the validator

adex001 committed:
feat-[166272333]: implement token upon signin or signup
implement feedback
update validation response
create a response helper for global errors

adex001 committed:
feat-[166272333]: implement token upon signin or signup
- resolve merge conflict from pulling from develop

adex001 committed:
feat-[166272333]: implement token upon signin or signup
- add more tests for user register and user login
- update swagger api to reflect changes made to the endpoint

adex001 committed:
feat-[166272333]: implement token upon signin or signup
- add more tests for user register and user login
- update swagger api to reflect changes made to the endpoint

adex001 committed:
feat-[166272333]: implement token upon signin or signup
- add dist folder to .gitignore file

fidelisojeah commented on .gitignore:
```diff
@@ -38,3 +38,4 @@ node_modules
 .env
 
 server/database_test.sqlite
+dist/
```
new line

adex001 replied:
alright..

adex001 committed:
feat-[166272333]: implement token upon signin or signup
add a new line after dist in .gitignore file

fidelisojeah commented on server/helpers/Token.js:
```diff
@@ -0,0 +1,48 @@
+   * @memberof Token
+   */
+  static async create(payload) {
+    const token = await jwt.sign(payload, process.env.SECRET, {
```
could you:

move `process.env. SECRET` and `process.env.TOKEN_EXPIRE` out of the class and set their values there:

```javascript
const tokenSecret = process.env.SECRET || 'secret';
const tokenExpiration = process.env.TOKEN_EXPIRE || '1d';
```

then here, you have 
```javascript
const token = await jwt.sign(payload, tokenSecret, { 
...
```

and in the verify token, you pick the values there as well.

We want a situation where the app does not break if someone forgets to include those values (and also not to always repeat ourself incase

adex001 replied:
alright. It has been implemented.

adex001 committed:
feat-[166272333]: implement token upon signin or signup
update token to have a default in case of no environment variable
update salt to have a default in case of no environment variable

despeauxz reviewed (commented):
nice job, just feel like some more test could be added to check some inputs like an invalid email format and both fields absent

adex001 commented:
> nice job, just feel like some more test could be added to check some inputs like an invalid email format and both fields absent

Thank you for your feedback. I will add more tests to the user.

adex001 committed:
feat-[166272333]: implement token upon signin or signup
resolve Merge conflict

adex001 committed:
feat-[166272333]: implement token upon signin or signup
remove console.log from user.login
update sinon as a dependency

adex001 committed:
feat-[166272333]: implement token upon signin or signup
remove console.log from user.login
update sinon as a dependency

adex001 committed:
Merge branch '166272333-ft-jwt-successful-registration' of https://github.com/andela/kingsmen-ah-backend into 166272333-ft-jwt-successful-registration

adex001 committed:
feat-[166272333]: implement token upon signin or signup
update package.json to include node and npm versions as engine

fidelisojeah commented on package.json:
```diff
@@ -57,8 +58,11 @@
+    "nyc": "^14.1.1"
+  },
+  "engines": {
+    "node": ">=6.0.0",
```
why node `6` not `10` ?

adex001 replied:
I wanted it to be backward compatible. That was why I used >= 6.

fidelisojeah commented on package.json:
```diff
@@ -57,8 +58,11 @@
+  },
+  "engines": {
+    "node": ">=6.0.0",
+    "npm": ">1.3.13"
```
why version `1.3` ? version `6` is deff stable

adex001 replied:
I wanted it to be backward compatible. Should I lock it to version 6 and above?

fidelisojeah reviewed (commented):
Please update the swagger doc (the input for sign up in swagger is invalid)

adex001 commented:
> Please update the swagger doc (the input for sign up in swagger is invalid)

okay. I got this..

adex001 committed:
feat-[166272333]: implement token upon signin or signup
update package.json to include node and npm version 10 and 6 respectively
update swagger documentation for register endpoint

adex001 committed:
feat-[166272333]: implement token upon signin or signup
update package.json to include node and npm version 10 and 6 for heroku

despeauxz reviewed (commented):
Good job, the API docs expects the first name as a required entity which isn't so in the endpoint. Please, can you update the swagger docs to reflect that?

adex001 commented:
> Good job, the API docs expects the first name as a required entity which isn't so in the endpoint. Please, can you update the swagger docs to reflect that?

alright! Thanks.

adex001 committed:
feat-[166272333]: implement token upon signin or signup
update swagger to remove firstname required status
